### PR TITLE
Add new problem: Mixed Precision Training

### PR DESCRIPTION
**Pull request by komaksym**
Resolves: #480

### CODE REVIEW

File: questions/161_mixed_precision_training/tests.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    "output": "Loss: 665.0000\nLoss dtype: float\nGradients: [0.5 -0.25]\nGrad dtype: float32"
+  },
+  {
+    "input": "import numpy as np\nmp = MixedPrecision(loss_scale=512.0)\nweights = np.array([1.0, 0.5], dtype=np.float64)\ninputs = np.array([[2.0, 1.0]], dtype=np.float64)\ntargets = np.array([3.0], dtype=np.float64)\nloss = mp.forward(weights, inputs, targets)\nprint(f\"Loss: {loss:.1f}\")\nprint(f\"Loss dtype: {type(loss).__name__}\")\ngrads = np.array([1024.0, 512.0], dtype=np.float16)\nresult = mp.backward(grads)\nprint(f\"Gradients: [{result[0]:.0f} {result[1]:.0f}]\")\nprint(f\"Grad dtype: {result.dtype}\")",
+    "output": "Loss: 128.0\nLoss dtype: float\nGradients: [2 1]\nGrad dtype: float32"
```
Comment (moe18) [inline questions/161_mixed_precision_training/tests.json:8]: the test needs to be simpler for the current way the site works, for would we be able to split up the test cases like this 
 {
            "test": "import numpy as np\nmp = MixedPrecision(loss_scale=1024.0)\nweights = np.array([0.5, -0.3], dtype=np.float32)\ninputs = np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32)\ntargets = np.array([1.0, 0.0], dtype=np.float32)\nloss = mp.forward(weights, inputs, targets)\ngrads = np.array([1024.0, 512.0], dtype=np.float16)\n\nresult = mp.backward(grads)\nprint(result)",
            "expected_output": "[1., 0.5]"
        },
        {
            "test": "import numpy as np\nmp = MixedPrecision(loss_scale=1024.0)\nweights = np.array([0.5, -0.3], dtype=np.float32)\ninputs = np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32)\ntargets = np.array([1.0, 0.0], dtype=np.float32)\nloss = mp.forward(weights, inputs, targets)\ngrads = np.array([1024.0, 512.0], dtype=np.float16)\n\nresult = mp.backward(grads)\nprint(loss)",
            "expected_output": "665.0"
        }

**Review by moe18 — COMMENTED** [review]
had some comments on the tests

File: questions/161_mixed_precision_training/tests.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+[
+  {
+    "input": "import numpy as np\nmp = MixedPrecision(loss_scale=1024.0)\nweights = np.array([0.5, -0.3], dtype=np.float32)\ninputs = np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32)\ntargets = np.array([1.0, 0.0], dtype=np.float32)\nloss = mp.forward(weights, inputs, targets)\nprint(f\"Loss: {loss:.4f}\")\nprint(f\"Loss dtype: {type(loss).__name__}\")\ngrads = np.array([512.0, -256.0], dtype=np.float32)\nresult = mp.backward(grads)\nprint(f\"Gradients: {result}\")\nprint(f\"Grad dtype: {result.dtype}\")",
```
[the remark this answers is not in the document]
Comment (komaksym) [reply questions/161_mixed_precision_training/tests.json]: Fixed

File: questions/160_mixed_precision_training/tests.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
+[
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=1024.0)\nweights = np.array([0.5, -0.3], dtype=np.float32)\ninputs = np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32)\ntargets = np.array([1.0, 0.0], dtype=np.float32)\nloss = mp.forward(weights, inputs, targets)\nprint(f\"Loss: {loss:.4f}\")\nprint(f\"Loss dtype: {type(loss).__name__}\")",
+        "expected_output": "Loss: 665.0000\nLoss dtype: float32"
```
Comment (moe18) [inline questions/160_mixed_precision_training/tests.json:4]: it seems like using type(loss).__name__ gives `float` not `float32`
Loss: 665.0000 Loss dtype: float

File: questions/160_mixed_precision_training/tests.json [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=1024.0)\ngrads = np.array([512.0, -256.0], dtype=np.float32)\nresult = mp.backward(grads)\nprint(f\"Gradients: {result}\")\nprint(f\"Grad dtype: {result.dtype}\")",
+        "expected_output": "Gradients: [ 0.5  -0.25]\nGrad dtype: float32"
+    },
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=512.0)\nweights = np.array([1.0, 0.5], dtype=np.float64)\ninputs = np.array([[2.0, 1.0]], dtype=np.float64)\ntargets = np.array([3.0], dtype=np.float64)\nloss = mp.forward(weights, inputs, targets)\nprint(f\"Loss: {loss:.1f}\")\nprint(f\"Loss dtype: {type(loss).__name__}\")",
+        "expected_output": "Loss: 128.0\nLoss dtype: float32"
```
Comment (moe18) [inline questions/160_mixed_precision_training/tests.json:12]: should be `float` not `float32`

File: questions/160_mixed_precision_training/tests.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=512.0)\ngrads = np.array([1024.0, 512.0], dtype=np.float16)\nresult = mp.backward(grads)\nprint(f\"Gradients: [{result[0]:.0f} {result[1]:.0f}]\")\nprint(f\"Grad dtype: {result.dtype}\")",
+        "expected_output": "Gradients: [2 1]\nGrad dtype: float32"
+    },
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=100.0)\nweights = np.array([0.1, 0.2], dtype=np.float32)\ninputs = np.array([[1.0, 1.0]], dtype=np.float32)\ntargets = np.array([0.5], dtype=np.float32)\nloss = mp.forward(weights, inputs, targets)\nprint(f\"Loss: {loss:.1f}\")\nprint(f\"Loss dtype: {type(loss).__name__}\")",
+        "expected_output": "Loss: 4.0\nLoss dtype: float32"
```
Comment (moe18) [inline questions/160_mixed_precision_training/tests.json:20]: see above

File: questions/160_mixed_precision_training/tests.json [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=100.0)\ngrads = np.array([200.0, 100.0], dtype=np.float64)\nresult = mp.backward(grads)\nprint(f\"Gradients: [{result[0]:.0f} {result[1]:.0f}]\")\nprint(f\"Grad dtype: {result.dtype}\")",
+        "expected_output": "Gradients: [2 1]\nGrad dtype: float32"
+    },
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=2048.0)\nweights = np.array([0.25], dtype=np.float64)\ninputs = np.array([[4.0]], dtype=np.float64)\ntargets = np.array([2.0], dtype=np.float64)\nloss = mp.forward(weights, inputs, targets)\nprint(f\"Loss: {loss:.1f}\")\nprint(f\"Loss dtype: {type(loss).__name__}\")",
+        "expected_output": "Loss: 2048.0\nLoss dtype: float32"
```
Comment (moe18) [inline questions/160_mixed_precision_training/tests.json:28]: see above

File: questions/160_mixed_precision_training/tests.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=2048.0)\ngrads = np.array([np.nan], dtype=np.float16)\nresult = mp.backward(grads)\nprint(f\"Gradients: [{result[0]:.0f}]\")\nprint(f\"Grad dtype: {result.dtype}\")",
+        "expected_output": "Gradients: [0]\nGrad dtype: float32"
+    },
+    {
+        "test": "import numpy as np\nmp = MixedPrecision(loss_scale=256.0)\nweights = np.array([1.0], dtype=np.float16)\ninputs = np.array([[2.0]], dtype=np.float16)\ntargets = np.array([3.0], dtype=np.float16)\nloss = mp.forward(weights, inputs, targets)\nprint(f\"Loss: {loss:.1f}\")\nprint(f\"Loss dtype: {type(loss).__name__}\")",
+        "expected_output": "Loss: 256.0\nLoss dtype: float32"
```
Comment (moe18) [inline questions/160_mixed_precision_training/tests.json:36]: ^

**Review by moe18 — COMMENTED** [review]
the test look very good, thanks for the changes, there was an issue with some of the tests check out the comments

**Comment by komaksym** [comment]
Thanks for the review, @moe18. I'll fix this next month when I'll have access to a computer

File: questions/160_mixed_precision_training/tests.json (hunk 3 above)
Comment (komaksym) [reply questions/160_mixed_precision_training/tests.json:4]: Still, it's float32 on my end

File: questions/160_mixed_precision_training/tests.json (hunk 4 above)
Comment (komaksym) [reply questions/160_mixed_precision_training/tests.json:12]: It's `float32` for me, as expected

**Comment by komaksym** [comment]
Even though for me the test cases were as expected, I still updated them to match your testing env @moe18
